src/bridge/cockpitpcpmetrics.c

```diff
@@ -675,6 +760,255 @@ cockpit_pcp_metrics_prepare (CockpitChannel *channel)
         }
     }
 
+  result = TRUE;
+
+ out:
+  g_free (instances);
+  g_free (omit_instances);
+  return result;
+}
+
+static void start_archive (CockpitPcpMetrics *self, gint64 timestamp);
+
+static gboolean
+add_archive (CockpitPcpMetrics *self,
+             const gchar *name)
+{
+  ArchiveInfo *info;
+  pmLogLabel label;
+  int rc;
+
+  info = g_new0 (ArchiveInfo, 1);
+  info->context = pmNewContext (PM_CONTEXT_ARCHIVE, name);
+  if (info->context < 0)
+    {
+      g_message ("%s: couldn't create pcp archive context for %s: %s", self->name, name, pmErrStr (info->context));
+      g_free (info);
+      return FALSE;
+    }
+
+  rc = pmGetArchiveLabel (&label);
+  if (rc < 0)
+    {
+      g_message ("%s: couldn't read archive label of %s: %s", self->name, name, pmErrStr (rc));
+      pmDestroyContext (info->context);
+      g_free (info);
+      return FALSE;
+    }
+
+  info->start = label.ll_start.tv_sec*1000 + label.ll_start.tv_usec / 1000;
+  self->archives = g_list_prepend (self->archives, info);
+  return TRUE;
+}
+
+static gint
+cmp_archive_start (gconstpointer a,
+                   gconstpointer b)
+{
+  const ArchiveInfo *a_info = a;
+  const ArchiveInfo *b_info = b;
+
+  if (a_info->start > b_info->start)
+    return 1;
+  else if (a_info->start < b_info->start)
+    return -1;
+  else
+    return 0;
+}
+
+static gboolean
+prepare_archives (CockpitPcpMetrics *self,
+                  const gchar *name,
+                  gint64 timestamp)
+{
+  GDir *dir;
+
+  dir = g_dir_open (name, 0, NULL);
+  if (dir)
+    {
+      const gchar *entry;
+      while ((entry = g_dir_read_name (dir)))
+        {
+          if (g_str_has_suffix (entry, ".meta"))
+            {
+              gchar *path = g_build_filename (name, entry, NULL);
+              path[strlen(path)-strlen(".meta")] = '\0';
+              add_archive (self, path);
+              g_free (path);
+            }
+        }
+      g_dir_close (dir);
+    }
+  else
+    add_archive (self, name);
```

Comment:
Hmm, yes, agreed.  But we should show any error except ENOENT: When "/foo" exists, it should be a directory of archives.  If it doesn't exist, "/foo.meta" should exist instead.